cpp/src/arrow/device.h

```diff
@@ -119,13 +119,16 @@ class ARROW_EXPORT MemoryManager : public std::enable_shared_from_this<MemoryMan
   /// The buffer will be allocated in the device's memory.
   virtual Result<std::unique_ptr<Buffer>> AllocateBuffer(int64_t size) = 0;
 
-  // XXX Should this take a `const Buffer&` instead
   /// \brief Copy a Buffer to a destination MemoryManager
   ///
   /// See also the Buffer::Copy shorthand.
   static Result<std::shared_ptr<Buffer>> CopyBuffer(
       const std::shared_ptr<Buffer>& source, const std::shared_ptr<MemoryManager>& to);
 
+  /// \brief Copy a Buffer to a destination MemoryManager
+  static Result<std::unique_ptr<Buffer>> CopyBuffer(
+      const Buffer& source, const std::shared_ptr<MemoryManager>& to);
```

Comment:
I would recommend choosing a different signature and perhaps a different name to avoid programming errors, e.g.:
```c++
  // \brief Copy a non-owned memory area to a destination MemoryManager
  static Result<std::unique_ptr<Buffer>> CopyMemory(
      uintptr_t source_address, int64_t size, const std::shared_ptr<MemoryManager>& to);
```